LandSandBoat/server
# [LUA] Create job_utils/dark_knight and audit JA functionality

**_I affirm:_**

## What does this pull request do?



**NOTE:** These lists will be updated as progress is made on the PR.

**Job Abilities**
- Create job_utils/dark_knight.lua and populates necessary ability check and ability use functions.
- Moves functionality from all Dark Knight globals/abilities files to globals/job_utils/dark_knight.lua.
- Adjusts Arcane Circle duration handling and adds consideration for new ARCANE_CIRCLE_POTENCY mod.
- Sets up Arcane Crest to be usable against Arcana enemies only.
- Sets up Blood Weapon to add duration from the Enhances "Blood Weapon" effect mod.
- Sets up Consume Mana to apply during weaponskills.
- Adds Dark Seal merit and augment tracking to globals/job_utils/dark_knight.lua.
- Sets up Diabolic Eye to apply merits to power and ENHANCES_DIABOLIC_EYE modifier to duration.
- Adds LAST_RESORT_DURATION modifier to Last Resort duration calculation.
- Sets up Last Resort to use DESPERATE_BLOWS merits to calculate the ENHANCES_DESPERATE_BLOWS bonus.
- Sets up Nether Void to apply Job Points and ENHANCES_NETHER_VOID modifier.
- Sets up Nether Void handling in new spells global absorb_spell.lua.
- Sets up Scarlet Delirium Job Points and adds them to Scarlet Delirium duration.
- Sets up Souleater to use MUTED_SOULS merits to calculate the ENHANCES_MUTED_SOUL bonus.
- Sets up Weapon Bash to inflict a Chainbound status when the player is equipped with Ignominy Gauntlets +2/+3.

**Effects**
- Sets up the Arcane Crest status effect and debuffs.
- Sets up Dark Seal effect DARK_MAGIC_CAST and DARK_MAGIC_DURATION modifiers.
- Removes base modifier from Diabolic Eye effect.
- Sets up Soul Enslavement effect ENSPELL and ENSPELL_DMG modifiers.
- Adds handling for Last Resort effect Subtle Blow bonus from ENHANCES_DESPERATE_BLOWS modifier.
- Adds handling for Souleater effect Zanshin bonus from ENHANCES_MUTED_SOUL modifier.

**Spells**
- Adds magic.lua calculateDuration calls to Bio, Bio II, Dread Spikes, Endark, Endark II, and Stun for DARK_MAGIC_DURATION modifier.
- Updates all Absorb-Stat (ACC/STR/DEX/etc) spells with unified formatting and power calculations.
- Updates Drain II and Drain III Max HP Boost durations.
- Updates Absorb-TP formatting, power, and ENHANCES_ABSORB_TP modifier.
- Creates Absorb-Attri.
- Creates Drain III.
- Creates Endark II.
- Creates Aspir III.

**Mods**
- Converts ARCANE_CIRCLE_DURATION from a 90s increase to a 50% increase on Chaos/Ignominy Sollerets.
- Creates ARCANE_CIRCLE_POTENCY and adds it to ability base strength on Chaos/Ignominy Sollerets.
- Creates ENHANCES_BLOOD_WEAPON and adds it to Blood Weapon base duration.
- Creates DARK_MAGIC_CAST and adds it to Dark Seal effect.
- Creates DARK_MAGIC_DURATION and adds it to Dark Seal effect.
- Creates ENHANCES_DARK_SEAL and adds it to Dark Seal subPower.
- Creates ENHANCES_ABSORB_TP and adds it to Absorb TP drain calculation.
- Creates ENHANCES_DIABOLIC_EYE and adds it to Diabolic Eye duration.
- Creates LAST_RESORT_DURATION and adds it to Last Resort duration.
- Creates ABSORB_DURATION and adds TODO to implement in magic.lua calculateDuration function.
- Creates ENHANCES_NETHER_VOID and adds it to Nether Void effect.
- Creates SOULEATER_EFFECT_POTENCY and adds it to the Enhances "Souleater" effect augment.
- Creates ENHANCES_MUTED_SOUL and adds it to Souleater.
- Creates ENHANCES_DESPERATE_BLOWS and adds it to Last Resort.
- Creates WEAPON_BASH_CHAINBOUND and adds it to Weapon Bash.

**Core/LUA Functions**
- Creates consumeManaBonus function in globals/weaponskills.lua
- Creates scarletDeliriumBonus function in globals/weaponskills.lua.
- Adjusts souleaterBonus function in globals/weaponskills.lua.
- Removes Dark Seal consideration from globals/spells/damage_spell.lua.
- Creates calculateScarletDeliriumMultiplier function in globals/spells/damage_spell.lua.
- Creates globals/spells/absorb_spell.lua and moves all handling for Absorb-Stat/Attri/TP/Drain/Aspir spells into the new global.
- Adds DARK_MAGIC_CAST handling to battleutils.cpp CalculateSpellCastTime function.
- Adds handling for removing Dark Seal after a successful Dark Magic spell cast to magic_state.cpp ApplyEnmity function.
- Creates HandleScarletDelirium function in battleutils.cpp.
- Adds Soul Enslavement Drain TP handling to HandleEnspell function in battleutils.cpp.
- Updates doSoulEaterEffect function in battleutils.cpp.
- Adds Scarlet Delirium damage bonus handling to ProcessDamage function in attack.cpp.
- Creates ENSPELL_SOUL_ENSLAVEMENT in ENSPELL enum in battleutils.h.
- Creates getMaxGearMod function charutils.cpp and charutils.h.

**Database**
- Adjusts ARCANE_CIRCLE_DURATION from +duration to * %duration on Chaos/Ignominy Sollerets in item_mods.sql.
- Adds ARCANE_CIRCLE_POTENCY to Chaos/Ignominy Sollerets in item_mods.sql.
- Adds ENHANCES_ABSORB_TP to Bale/Heathen's Gauntlets in item_mods.sql.
- Adds ENHANCES_NETHER_VOID to Bale/Heathen's Flanchard in item_mods.sql
- Adds all modifiers to the Ankou's Mantle in item_mods.sql.
- Adds all modifiers to the Brutality in item_mods.sql.
- Adds all modifiers to the Agwu's Scythe in item_mods.sql.
- Adds WEAPON_BASH_CHAINBOUND to Ignominy Gauntlets +2/+3 in item_mods.sql.
- Update Last Resort latent description for Abyss/Fallen's Sollerets in item_latents.sql.
- Add Last Resort latent to Fallen's Sollerets +2/+3 in item_latents.sql.
- Adds ENHANCES_BLOOD_WEAPON to augment 1287 in augments.sql.
- Adds ENHANCES_DARK_SEAL to augment 1370 in augments.sql.
- Adds ENHANCES_DIABOLIC_EYE to augment 1371 in augments.sql.
- Adds SOULEATER_EFFECT_POTENCY to augment 897 in augments.sql
- Adds ENHANCES_MUTED_SOUL to augment 1372 in augments.sql.
- Adds ENHANCES_DESPERATE_BLOWS to augment 1373 in augments.
- Adjusts Dark Seal merit values in merits.sql.
- Removes EFFECTFLAG_MAGIC_END from Dark Seal in status_effects.sql.
- Adds Absorb-Attri to spell_list.sql.
- Adds Endark II to spell_list.sql.
- Adds Drain III to spell_list.sql.

## References

All job abilities - https://www.bg-wiki.com/ffxi/Dark_Knight

## Steps to test these changes


1. Upload changes to your server.
2. Load the server.
3. Use the `!changejob DRK 99` command.
4. Use all JA's to confirm they do not throw errors.
5. Test all JA's and compare their behavior to BGWiki JA descriptions.
